src/transport/CryptoContext.cpp

```diff
@@ -181,19 +185,30 @@ CHIP_ERROR CryptoContext::Encrypt(const uint8_t * input, size_t input_length, ui
     ReturnErrorOnFailure(GetIV(header, IV, sizeof(IV)));
     ReturnErrorOnFailure(GetAdditionalAuthData(header, AAD, aadLen));
 
-    KeyUsage usage = kR2IKey;
+    if (mKeyContext)
+    {
+        MutableByteSpan plaintext(output, input_length); // WARNING: ASSUMES input == output
+        MutableByteSpan mic(tag, taglen);
 
-    // Message is encrypted before sending. If the secure session was created by session
-    // initiator, we'll use I2R key to encrypt the message that's being transmitted.
-    // Otherwise, we'll use R2I key, as the responder is sending the message.
-    if (mSessionRole == SessionRole::kInitiator)
+        ReturnErrorOnFailure(mKeyContext->EncryptMessage(plaintext, ByteSpan(AAD, aadLen), ByteSpan(IV, sizeof(IV)), mic));
+    }
+    else
     {
-        usage = kI2RKey;
+        VerifyOrReturnError(mKeyAvailable, CHIP_ERROR_INVALID_USE_OF_SESSION_KEY);
+        KeyUsage usage = kR2IKey;
+
+        // Message is encrypted before sending. If the secure session was created by session
+        // initiator, we'll use I2R key to encrypt the message that's being transmitted.
+        // Otherwise, we'll use R2I key, as the responder is sending the message.
+        if (mSessionRole == SessionRole::kInitiator)
+        {
+            usage = kI2RKey;
+        }
+
+        ReturnErrorOnFailure(AES_CCM_encrypt(input, input_length, AAD, aadLen, mKeys[usage], Crypto::kAES_CCM128_Key_Length, IV,
+                                             sizeof(IV), output, tag, taglen));
```

Comment:
Hopefully we can refactor the code to make use of KeyContext for PASE/CASE as well, not just groups

Reply:
That's the idea. Hopefully, the unicast symmetric encryption will be refactored to use SymmetricKeyContex too, so we can remove the key value from CryptoContext.